### ISSUE TITLE
[BUG][QUESTION] cannot download private video

### ISSUE THREAD

**Issue by sokol8** [pytube/pytube#1097]
**Before creating an issue**
pytube 11.0.0 here

**Describe the bug**

I'm trying to download a private video which I can see as a logged in user.

I'm using  the following command in Python3
>>> YouTube('https://www.youtube.com/watch?v=RUFKT3N5ow8', use_oauth=True).streams.first().download()

which yields 

pytube.exceptions.VideoPrivate: RUFKT3N5ow8 is a private video

**To Reproduce**
 >>> from pytube import YouTube
 >>> YouTube('https://www.youtube.com/watch?v=RUFKT3N5ow8', use_oauth=True).streams.first().download()

**Expected behavior**
I'm expecting some way to authenticate with Google and authorize the pytube client so I can download the video using my google account and pytube.

I was checking the code of innertube and I can see a place where the token file is supposed to be read
if os.path.exists(_token_file):

and where self.fetch_bearer_token() call is made but I cannot figure out how to make it work for me.

I can see a place in 
def _call_api(self, endpoint, query, data):

where self.fetch_bearer_token() but I cannot see out of the code why fetch_bearer_token is not producing output it's supposed to produce.

I'm happy to help extending the code to improve support if needed. Just need some guidance in the start.

**Output**
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 284, in streams
    self.check_availability()
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 217, in check_availability
    raise exceptions.VideoPrivate(video_id=self.video_id)
pytube.exceptions.VideoPrivate: RUFKT3N5ow8 is a private video

**System information**
Please provide the following information:

- Python version (run `python --version`)
$python3 --version
Python 3.9.5

- Pytube version (run `print(pytube.__version__)` in python)
$ pytube --version
pytube 11.0.0

- Command used to install pytube
pip3 install pytube

**Comment by sokol8** [comment]
@89z my question is more about how to use OAuth with pytube.

I was trying to download an Unlisted video on youtube which any signed-in Google user could see. 

I can see that pytube has some built in support to Authenticate and Authorize with Google but I cannot see how to make it work. 

** Is this feature available already? **
I searched hard through documentation and I could not find any reference to it.

I can see code inside fetch_bearer_token(self) function but it doesn't seem to run at all.
at least I cannot see results of execution of 

print(f'Please open {verification_url} and input code {user_code}')
input('Press enter when you have completed this step.')

**Comment by sokol8** [comment]
@89z you are correct - the video is not available in the browser either anymore

I just uploaded a private video to youtube myself to test

Here's the video https://www.youtube.com/watch?v=PErDRwl1Tkc

Here's the code

>>> from pytube import YouTube
>>> YouTube('https://www.youtube.com/watch?v=PErDRwl1Tkc').streams.first().download()


here's the result

>>> YouTube('https://www.youtube.com/watch?v=PErDRwl1Tkc', use_oauth=True).streams.first().download()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 284, in streams
    self.check_availability()
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 217, in check_availability
    raise exceptions.VideoPrivate(video_id=self.video_id)
pytube.exceptions.VideoPrivate: PErDRwl1Tkc is a private video

**Comment by sokol8** [comment]
@89z if you provide me your gmail I will add you to the list

**Comment by sokol8** [comment]
@89z just added you to the access list

**Comment by sokol8** [comment]
what am I doing wrong then?

```
python3
Python 3.9.5 (default, May  4 2021, 03:36:27) 
[Clang 12.0.0 (clang-1200.0.32.29)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import pytube
>>> print(pytube.__version__)
11.0.0
>>> from pytube import YouTube
>>> YouTube('https://www.youtube.com/watch?v=PErDRwl1Tkc', use_oauth=True).streams.first().download()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 284, in streams
    self.check_availability()
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 217, in check_availability
    raise exceptions.VideoPrivate(video_id=self.video_id)
pytube.exceptions.VideoPrivate: PErDRwl1Tkc is a private video
```

**Comment by sokol8** [comment]
If I start fetching a title - I get to the Authentication prompt. But not when I start download

```
>>> yt = YouTube('https://www.youtube.com/watch?v=RUFKT3N5ow8', use_oauth=True)
>>> print(yt.title())
Please open https://www.google.com/device and input code JWD-BDV-HGF
Press enter when you have completed this step.

```

**Comment by sokol8** [comment]
and after I did the whole authentication and authorization steps - bummer

```
yt = YouTube('https://www.youtube.com/watch?v=RUFKT3N5ow8', use_oauth=True)
>>> print(yt.title())
Please open https://www.google.com/device and input code JWD-BDV-HGF
Press enter when you have completed this step.
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 330, in title
    self._title = self.vid_info['videoDetails']['title']
KeyError: 'videoDetails'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 334, in title
    self.check_availability()
  File "/usr/local/lib/python3.9/site-packages/pytube/__main__.py", line 217, in check_availability
    raise exceptions.VideoPrivate(video_id=self.video_id)
pytube.exceptions.VideoPrivate: RUFKT3N5ow8 is a private video
```

**Comment by sokol8** [comment]
@89z first of all -  thanks for helping me out with this! I'm an occasional user of pytube for the last 3 years or so but I love the tool and really happy to see it being maintained and developed. 

Back to the code - all I tried to do was

```
>>> from pytube import YouTube
>>> YouTube('https://www.youtube.com/watch?v=PErDRwl1Tkc', use_oauth=True).streams.first().download()
```

so I am not sure what can be erroneous on my side of using the code.
the line 
```
YouTube('https://www.youtube.com/watch?v=PErDRwl1Tkc', use_oauth=True).streams.first().download() 
```
is taken from the official pytube documentation

**Comment by tfdahlin** [comment]
@89z, I appreciate that you're trying to help, but people are asking for solutions to problems with pytube. Please keep your responses within that context.

@sokol8 give me a few to see if I can figure out what's going on here. I thought I thoroughly tested this, but I may have missed something silly.

**Comment by 89z** [comment]
@tfdahlin some comments have been removed. I went back and forth with OP several times before suggesting another package.

**Comment by tfdahlin** [comment]
@sokol8 figured out the problem -- before the OAuth implementation, I had some early checks to see if a video was unavailable that are no longer an accurate representation of our capabilities. I'll fix that.

**Comment by tfdahlin** [comment]
Try out the branch I just made, and let me know if it works for you:
`python3 -m pip install git@https://github.com/tfdahlin/pytube@fix/1097`

It works for the private video I tested against, but would appreciate extra eyes. If it fixes your problem, I'll merge it this weekend.

**Comment by tfdahlin** [comment]
> 
> 
> ** Is this feature available already? **
> I searched hard through documentation and I could not find any reference to it.
> 
> I can see code inside fetch_bearer_token(self) function but it doesn't seem to run at all.
> at least I cannot see results of execution of
> 
> print(f'Please open {verification_url} and input code {user_code}')
> input('Press enter when you have completed this step.')

My apologies about the lack of documentation. This was something that I started a couple weeks ago, but didn't get around to merging until today. The documentation I added isn't the best, but there's at least some mention of how it's supposed to work at the [end of the quickstart section](https://pytube.io/en/latest/user/quickstart.html) now.

**Comment by 89z** [comment]
@tfdahlin I thought about it, and I wanted you to know that Im taking PyTube off the list here:

https://github.com/89z/mech/tree/master/youtube#other-repos

which means I wont be monitoring issues here anymore. Maybe you dont care, but I thought I at least owed you the courtesy to tell you, given what I consider positive past interactions. Youre one of the few people on GitHub ive worked with thats been patient, and actually thought critically about issues, instead of slamming them shut without much thought. That makes for better software, so I appreciate that. However I dont appreciate your tone of the comment above. I earnestly tried to help this guy, and it was only after exhausting those paths that I jumped to my own library.

Take care.

**Comment by glubsy** [comment]
@89z nah, don't take it the wrong way. It's pretty clear you were trying to help OP. You didn't have to delete your comments for so much.

**Comment by tfdahlin** [comment]
@89z I also had next to zero context besides sokol's comments when I came in with my responses. I appreciate you trying to help, but from my perspective when I arrived, all that I saw was "go use my library instead", which isn't very helpful when somebody may be trying to e.g. write a python application for interacting with youtube.

One of the core reasons that people choose pytube over youtube-dl for example is because of our API design that makes it relatively easy to work with. When somebody asks for help using it, and are told "use a different library in another language", that isn't really helpful for the problem that they're having. Your comments elsewhere have been immensely helpful (especially re: implementing OAuth), but when you delete them shortly after being done with the conversation makes it very hard for myself and others to understand context or reference the useful information later.

**Comment by 89z** [comment]
> but when you delete them shortly after being done with the conversation makes it very hard for myself and others to understand context or reference the useful information later.

Sorry, but I cant stop doing that. I dont like the practice, but too many people like to look at past conversations in negative light, and dont want to consider factors that were important at the time. Because of this, Ive been banned from the Golang project on GitHub, among other issues. So its just a means to protect myself from shitty people who look on after the fact (everyone in this thread is cool, except that one comment I am talking about).

**Comment by sokol8** [comment]
hi guys
@tfdahlin great thanks for the fix. Will give it a try shortly. 

In all sincerity, @89z was trying to help and asked me very good questions which helped us isolate the problem (in a sense that we got to a properly setup private test video which definitely didn't work with pytube at the time of discussion). 

I believe @89z could not solve the issue in the code, and recommended a solution which would actually solve my immediate problem of being unable to download a Private Video from CLI. When he mentioned his own solution there was also a clear remark that it's not in Python.

@89z great thanks for the help! - I saw your mech repo when checking out your profile. Will give it a try as well. I've just been a fan of pytube for years though never used it API - just occasional CLI download. 
I feel sorry you're leaving this repo because of sincerely trying to help me.

**Comment by BabbarRaghav** [comment]
> Try out the branch I just made, and let me know if it works for you: `python3 -m pip install git@https://github.com/tfdahlin/pytube@fix/1097`
> 
> It works for the private video I tested against, but would appreciate extra eyes. If it fixes your problem, I'll merge it this weekend.

i am not able to install this fix

**Comment by BabbarRaghav** [comment]
> hi guys @tfdahlin great thanks for the fix. Will give it a try shortly.

hey @sokol8 , is your problem solved???

**Comment by tiff1002** [comment]
I have managed to get this working but I had to uninstall pytube then install the branch with:

python -m pip install git+https://github.com/tfdahlin/pytube.git@fix/1097

then I had to apply a manual fix to cipher.py from:

https://github.com/pytube/pytube/issues/1243

then the following worked (had to interact with console output oauth the first time but it seems to cache):

test = pytube.YouTube(private_video_url, use_oauth=True)

test.streams.order_by('resolution').desc().first().download(filename_prefix="video_")


but things do not seem to work from your 12.0 main branch can we get the private video fix and the fix to cipher.py from https://github.com/pytube/pytube/issues/1243 merged into the master branch so that I can use the latest version with these fixes pretty please

**Comment by tiff1002** [comment]
> Try out the branch I just made, and let me know if it works for you: `python3 -m pip install git@https://github.com/tfdahlin/pytube@fix/1097`
> 
> It works for the private video I tested against, but would appreciate extra eyes. If it fixes your problem, I'll merge it this weekend.

see my last reply.  How can we get this fix and this one https://github.com/pytube/pytube/issues/1243

into the core project?

**Comment by Claire874** [comment]
Hi guys!
I have the same problem, when I want to download the dataset. I got:
`pytube.exceptions.VideoPrivate: wX78EPtSuzU is a private video
`
And I tried the method mentioned before, but cannot solve my bug.
